chcemvediet/apps/accounts/tests/test_views.py

```diff
@@ -25,9 +26,11 @@ def test_allowed_http_methods(self):
         allowed = [u'HEAD', u'GET']
         self.assert_allowed_http_methods(allowed, reverse(u'accounts:profile'))
 
-    @unittest.skip(u'FIXME')
     def test_anonymous_user_is_redirected(self):
-        self.assert_anonymous_user_is_redirected(reverse(u'accounts:profile'))
+        patched = functools.partial(page_not_found, template_name=u'404x.html')
```

Comment:
@martinmacko47 Zvolil som tento sposob, lebo z `TEMPLATE_LOADERS` potrebujem aj template `accounts/login.html`.

Reply:
@viliambalaz Tento test uz netestuje vseobecnu libku z adresara `poleno/`, ale cast samotnej aplikacie. Cize moze zavisiet na ostatnych castiach aplikacie. Cize nerozumiem preco redirect na `accounts/login/` skonci na 404 Not found. Mal by normalne skoncit s 200 OK. Skus zistit, preco ten redirect nefunguje normalne.

Jedna vec, co mi napada, je, ze mozno problem je v tom, ze ta URL, kam sa to redirectuje ma byt `en/accounts/login/` s locale na zaciatku.

Reply:
@viliambalaz Problem vyrieseny, vid odpoved v slacku.

Reply:
Odpoved zo slacku:

tak si vsimnes, ze to pada vo funkcii `change_lang`. Ta sa pouziva iba v template `chcemvediet/templates/main/snippets/header/language.html`. Ked som vnutro `change_lang` zakomentoval a vratil z neho iba prazdny string, tak test zbehol. Cize problem je vo vnutri `change_lang`. Konkretne problem je, ze `context[u'request'].path` vo vnutri `change_lang` pocas testov vrati cestu bez `/en` prefixu. Hoci mimo testov, to pekne vracia cestu s `/en` prefixom. Pridaj si do `change_lang` kontrolny vypis `print(context[u'request'].path`) a spust si test a potom si spust normalne aplikaciu a porovnaj vysledok.
Cize treba zistit, preco v testoch je `context[u'request'].path` bez prefixu, hoci pocas behu aplikacie nie.

Reply:
Odpoved zo slacku:

K tomu erroru. Ono to je cele magicke. Problem je, ze v produkcnom mode, v devel mode a v testoch sa inac handluju errory z templatov a neexistujuce URL, takze chyba sa prejavuje iba v testoch. V produkcii aj v devel mode je prekryta niecim inym.
Auth redirect funguje tak, ze najprv usera redirectne na `/accounts/login/?next=xxx`, az potom `LocaleMiddleware` tuto URL redirectne na `/en/accounts/login/?next=xxx`.
Pocas testu: Ked sa resolvuje `/accounts/login` (bez language prefixu), tak kedze tuto URL nemame definovanu Django najprv raisne `Http404` (tu: https://github.com/django/django/blob/1.7.4/django/core/handlers/base.py#L98). Potom na zaklade tento exception vyrendruje nasu 404 obrazovku (tu: https://github.com/django/django/blob/1.7.4/django/core/handlers/base.py#L150). A vyrendrovanie nasej 404 obrazovky padne na tom, ze `change_lang` pouzite v `snippets/header/language.html` template nedokaze spracovat URL bez language prefixu.
V devel mode (so settingom `DEBUG=True`): Ked sa resolvuje `/accounts/login` (bez language prefixu), tak Django opat na rovnakom mieste najprv raisne `Http404`. Ale kedze je zapnuty debug mod, tak namiesto nasej 404 vyrendruje svoju debug 404 (tu: https://github.com/django/django/blob/1.7.4/django/core/handlers/base.py#L146). Nasledne `LocaleMiddleware` tuto vyrendrovanu 404 zahodi a namiesto nej vrati redirect (tu: https://github.com/django/django/blob/1.7.4/django/core/handlers/base.py#L204 a tu: https://github.com/django/django/blob/1.7.4/django/middleware/locale.py#L51). Takze v devel mode sa chyba neprejavi, pretoze nas template `snippets/header/language.html` sa nepouzije a tak redirect na `/en/accounts/login/?next=xxx` zafunguje.
V produkcnom mode (so settingom `DEBUG=False`): Ked sa resolvuje `/accounts/login` (bez language prefixu), tak Django opat na rovnakom mieste najprv raisne `Http404`. Kedze nie je zapnuty debug mod, tak Django sa pokusi vyrendrovat nasu 404, ktora raisne exception v `snippets/header/language.html` z `change_lang`, rovnako ako pocas testov. Avsak Django, pri vypnutom debug mode, ma feature, ktore skryva exceptions vzniknute v templatoch. Ak pri rendrovani templatu nastane exception, Django exception zachyti a namiesto vyrendrovaneho templatu pouzije prazdny string. Takze nasa chyba v `change_lang` sa takto umlci, a nasa 404 sa vyrendruje, akurat s chybajucou castou, ktora mala byt vyrendrovana v `snippets/header/language.html`. Na konci opat `LocaleMiddleware` tuto vyrendrovanu 404 zahodi a namiesto nej vrati redirect.

Skus si zapnut debugger a odtrasovat, co sa v kode deje pocas testu a pocas volania endpointu z browsra.
Rovnaka chyba nenastava iba pri login redirecte, ale kedykolvek sa otvori URL bez language prefixu. Akurat symptom chyby je na produkcii aj v devel mode umlcany inymi vecami.
Fixnut to treba opravenim `change_lang`, aby fungovala, aj ked je prave otvorena URL bez language prefixu. Asi najlepsie tak, ze ak sa `context[u'request'].path` nepodari resolvnut, tak change_lang vrati povodny `context[u'request'].path` bez snahy zmenit language danej URL.